### PR TITLE
feat: automatic persisted queries

### PR DESCRIPTION
**Pull request by ccollie**
This is the beginnings of an `apollo-client` compatible automatic persisted queries plugin.

Spec: https://github.com/apollographql/apollo-link-persisted-queries#apollo-engine
Apollo: https://www.apollographql.com/docs/apollo-server/performance/apq/

 It should server as a means of determining the support needed from `graphql-ez`

### CODE REVIEW

**Comment by PabloSzx** [comment]
Hi @ccollie , thank you and nice work on this plugin, meanwhile you were working on this plugin I added a new feature that allows you to hook into the processRequest without needing to wrap the handleRequest, https://github.com/PabloSzx/graphql-ez/commit/4c3ed735a4032303fbd328e0bdb338a6afca4b6b, this plugin can be greatly simplified thanks to this, and you only need to do something like 

```ts
(ctx.preProcessRequest ||= []).push(({ requestOptions }) => {
  requestOptions.query;
  requestOptions.extensions;
});
```

**Review by PabloSzx — COMMENTED** [review]
@ccollie I did a rebase and aligned some types and minor logic, fixed the "GET" test, and also improved the startFastifyServer `.query`, all the tests could be cleaned with the new improved "query" (which accepts extensions and you can give it an empty query), but it also meant more changes that I thought could be better be done from you 👍

File: packages/plugin/automatic-persisted-queries/package.json [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  "scripts": {
+    "dev": "bob-esbuild watch",
+    "prepack": "bob-esbuild build",
+    "postpublish": "gh-release",
+    "test": "jest"
+  },
+  "dependencies": {
+    "tiny-lru": "^7.0.6",
+    "@envelop/core": "^1.0.3",
+    "@envelop/types": "^1.0.2",
```
Comment (PabloSzx) [inline packages/plugin/automatic-persisted-queries/package.json:38]: You should reuse the Envelop returned from graphql-ez, graphql-ez re-exports everything from Envelop

Comment (PabloSzx) [inline packages/plugin/automatic-persisted-queries/package.json:38]:
```suggestion
```

File: packages/plugin/automatic-persisted-queries/src/errors.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
+import { EnvelopError } from '@envelop/core';
```
Comment (PabloSzx) [inline packages/plugin/automatic-persisted-queries/src/errors.ts:1]:
```suggestion
import { EnvelopError } from 'graphql-ez';
```

File: packages/plugin/automatic-persisted-queries/package.json (hunk 1 above)
Comment (PabloSzx) [reply packages/plugin/automatic-persisted-queries/package.json:38]: this ensures that graphql-ez is always 100% aligned with Envelop

**Review by PabloSzx — APPROVED** [review]

**Comment by PabloSzx** [comment]
thank you @ccollie 💯

**Comment by ccollie** [comment]
@PabloSzx I noticed that headers set in `Response`s are not merged in the adapters. See
https://github.com/ccollie/graphql-ez/blob/39d727a1de7cb7a8061dba13ebb22556b2b7283c/packages/plugin/automatic-persisted-queries/src/errors.ts#L54

for an example why this is helpful.

**Comment by PabloSzx** [comment]
> @PabloSzx I noticed that headers set in `Response`s are not merged in the adapters. See
> https://github.com/ccollie/graphql-ez/blob/39d727a1de7cb7a8061dba13ebb22556b2b7283c/packages/plugin/automatic-persisted-queries/src/errors.ts#L54
> 
> for an example why this is helpful.

yeah, now I'm pushing a fix, it was this way because of batched queries, but I will simply merge all the headers obtained for the batched queries situation

**Pull request merged**